File: matlab/demos/jmm/demo_jmm_2_2_motif.m
%% demo_jmm_2_2_motif.m
% Analysis 2.2 (Online Supplement, Section 8.1): motif discovery in
% Acknowledgement from pitch alone.
%
% A demo of the Music Perception Toolbox reproducing the analysis from
% the JMM article; lightly edited from the article's own script. Data
% come from the jmm package (Acknowledgement read from a MIDI
% transcription you supply); the figures stay on screen unless
% SAVE_FIGURES is set.
%
% Analysis 2.2: data-driven motif discovery in Coltrane's Acknowledgement.
%
% The melody is read as one event per note carrying pitch. The recurring
% four-note cell is then recovered as a peak in the density of short
% interval patterns, without being supplied in advance. Two routes reach
% the same transposition-invariant cell.
%
% Differenced route. Pitch is first-differenced to melodic intervals,
% collapsing every transposition of a figure onto the same interval
% pattern. Three consecutive intervals are then bound into an ordered
% super-event: each is a point in a three-dimensional interval space, and
% the density over those points is a smoothed recurrence count. A figure
% stated k times contributes k near-coincident points, so its interval
% triple stands out as a local maximum. Reading the density at every
% observed triple and ranking turns motif discovery into peak finding.
%
% Relative route. The same cell is reached without differencing, by
% binding four consecutive pitches into an ordered super-event taken
% relative: the common transposition is removed, so transposed statements
% again coincide, and three degrees of freedom remain --- the dimension
% of the differenced route's interval triple. The two densities therefore
% describe the same object through different internal metrics. The
% differenced form treats consecutive intervals independently; the
% relative form couples them through the pitch they share. With the
% interval kernel set to sqrt(2) times the per-pitch kernel --- an
% interval being a difference of two pitches, which is the scaling
% differenceEvents applies of its own accord --- the two rank the motifs
% identically, and differ only in that coupling: a faint shear in the
% relative density's slice, absent from the differenced one.
%
% A relative density is read in translation-reduced coordinates: an
% r-tuple minus its first value, so a cell's coordinates are its
% cumulative intervals. Both the cells and the slice grid are written
% that way for the relative route below.
%
% Pre-MAET structure:
%
%     attribute    order  sigma              rel  per
%     ----------   -----  -----------------  ---  ---
%     dp           3      sqrt(2) * 0.15 st  no   no     differenced route
%     pitch        4      0.15 st            yes  no     relative route
%
%     Ordered (exch = 0) in both. Estimator: the density read at each cell.
%
% Data: jmm.acknowledgement (the solo, from your own MIDI transcription at
% data/AwakeningSolo.mid). Toolbox: preMaetFromAttrTable, differenceEvents,
% bindEvents, buildMaet, evalMaet, showPreMaet. Runtime: a few seconds.

% The demo folder is located from the toolbox root, and adding it puts
% the +jmm helper package in scope.
mptRoot = which('buildMaet');
if isempty(mptRoot)
    error('demoJmm:toolboxNotFound', ...
        ['The toolbox is not on the path. Add the matlab folder of the ' ...
         'Music Perception Toolbox, then run this demo again.']);
end
thisDir = fullfile(fileparts(mptRoot), 'demos', 'jmm');
addpath(thisDir);
clear mptRoot

% Set true to write the figures to a figures/ folder beside this script;
% false leaves them on screen only.
SAVE_FIGURES = false;

prevDefaults = mptDefaults('showHints', false);

SIGMA_PITCH = 0.15;        % semitones (15 cents): the per-pitch uncertainty
R_DIFF      = 3;           % bound interval triples (a four-note cell)
R_REL       = 4;           % bound pitch quadruples (the same cell)
TOP         = 6;           % motifs shown in the ranking
ALS         = [3 -3 5];    % +m3, -m3, +P4: the "A Love Supreme" cell

C_ALS   = [0.761 0.314 0.031];   % the recurring "A Love Supreme" cell
C_OTHER = [0.122 0.306 0.722];   % the surrounding recurring cells

% --- the melody as a pre-MAET, one event per note ---------------------------
notes = jmm.acknowledgement();
fprintf('%d notes; pitch range %.0f-%.0f (MIDI); span %.1f QN\n', ...
        height(notes), min(notes.pitch), max(notes.pitch), ...
        max(notes.onsetBeats));
melody = preMaetFromAttrTable(notes, 'attributes', { ...
    struct('column', 'pitch', 'name', 'pitch', 'sigma', SIGMA_PITCH)}, ...
    'time', 'beats', 'chords', 'separate', 'weights', 'ones');

% --- differenced route: intervals bound into ordered triples -----------------
% differenceEvents widens the kernel by sqrt(2) itself, so the bound
% pre-MAET already carries the interval width and buildMaet needs no sigma.
diffRoute = bindEvents(differenceEvents(melody, 1), R_DIFF, 'step', 1);
showPreMaet(diffRoute, 'maxEvents', 3);
[pd, ~, ~] = unpackPreMaet(diffRoute);
cellsD = pd{1};                                   % 3 x nCells intervals
densD = evalMaet(buildMaet(diffRoute, 'verbose', false), cellsD, ...
                 'verbose', false);
[classD, countD, meanD] = localRank(cellsD, densD);

% --- relative route: pitches bound into ordered relative quadruples ---------
relRoute = bindEvents(melody, R_REL, 'step', 1, 'relOuter', true);
showPreMaet(relRoute, 'maxEvents', 3);
[pr, ~, ~] = unpackPreMaet(relRoute);
cellsR = pr{1};                                   % 4 x nCells pitches
ivR = diff(cellsR, 1, 1);                         % its interval triples
densR = evalMaet(buildMaet(relRoute, 'verbose', false), cumsum(ivR, 1), ...
                 'verbose', false);
[classR, countR, meanR] = localRank(ivR, densR);

% --- report -----------------------------------------------------------------
fprintf('\ndifferenced: %d ordered interval triples, %d distinct classes\n', ...
        size(cellsD, 2), size(classD, 1));
fprintf('relative:    %d ordered pitch quadruples, %d distinct classes\n\n', ...
        size(cellsR, 2), size(classR, 1));
fprintf('%4s  %16s  %5s  %11s  %9s   contour\n', ...
        'rank', 'interval class', 'count', 'differenced', 'relative');
for i = 1:TOP
    k = classD(i, :);
    [~, j] = ismember(k, classR, 'rows');
    if isequal(k, ALS)
        mark = '  <- A Love Supreme cell';
    else
        mark = '';
    end
    fprintf('%4d  %16s  %5d  %11.1f  %9.1f   %s%s\n', i, ...
            localTriple(k), countD(i), meanD(i), meanR(j), ...
            localContour(k), mark);
end
same = isempty(setdiff(classD(1:TOP + 2, :), classR(1:TOP + 2, :), 'rows'));
fprintf('\ntop-%d interval-class set identical: %d\n', TOP + 2, same);
[~, rankD] = ismember(ALS, classD, 'rows');
[~, rankR] = ismember(ALS, classR, 'rows');
fprintf('A Love Supreme cell %s: rank %d (differenced), rank %d (relative)\n', ...
        localTriple(ALS), rankD, rankR);

% --- the (+3, i2, i3) slice of each density ---------------------------------
% The plane of cells sharing the leading motif's first interval. The grid
% is written as interval triples for the differenced route and as their
% cumulative sums for the relative one.
gridStep = SIGMA_PITCH * sqrt(2) / 3;
gridAxis = -7:gridStep:7;
[I2, I3] = meshgrid(gridAxis, gridAxis);
gridIv = [repmat(ALS(1), 1, numel(I2)); I2(:).'; I3(:).'];
sliceD = reshape(evalMaet(buildMaet(diffRoute, 'verbose', false), gridIv, ...
                          'verbose', false), size(I2));
sliceR = reshape(evalMaet(buildMaet(relRoute, 'verbose', false), ...
                          cumsum(gridIv, 1), 'verbose', false), size(I2));
[~, iD] = max(sliceD(:));  [rD, cD] = ind2sub(size(sliceD), iD);
[~, iR] = max(sliceR(:));  [rR, cR] = ind2sub(size(sliceR), iR);
fprintf('\nslice peak (i2, i3): differenced (%.2f, %.2f), relative (%.2f, %.2f)\n', ...
        gridAxis(cD), gridAxis(rD), gridAxis(cR), gridAxis(rR));

% --- figure -----------------------------------------------------------------
fig = figure('Position', [50 50 1500 1070], 'Color', 'w');
axB1 = axes('Parent', fig, 'Position', [0.10 0.60 0.33 0.32]);
axB2 = axes('Parent', fig, 'Position', [0.58 0.60 0.33 0.32]);
axS1 = axes('Parent', fig, 'Position', [0.10 0.07 0.28 0.40]);
axS2 = axes('Parent', fig, 'Position', [0.58 0.07 0.28 0.40]);

localBars(axB1, classD, countD, meanD, TOP, ALS, C_ALS, C_OTHER, ...
          'Differenced interval triples');
localBars(axB2, classR, countR, meanR, TOP, ALS, C_ALS, C_OTHER, ...
          'Relative pitch quadruples');
localSlice(axS1, gridAxis, sliceD, classD, meanD, TOP, ALS, C_ALS, ...
           'Density through (+3, i_2, i_3): differenced');
localSlice(axS2, gridAxis, sliceR, classR, meanR, TOP, ALS, C_ALS, ...
           'Density through (+3, i_2, i_3): relative');
annotation(fig, 'textbox', [0.05 0.94 0.9 0.05], 'String', ...
    'Coltrane, Acknowledgement: motif density', ...
    'HorizontalAlignment', 'center', 'EdgeColor', 'none', 'FontSize', 15);

figDir = fullfile(thisDir, 'figures');
if SAVE_FIGURES
    if ~exist(figDir, 'dir'), mkdir(figDir); end
    print(fig, '-dpng', '-r140', fullfile(figDir, 'demo_jmm_2_2_motif.png'));
    fprintf('Saved figures/demo_jmm_2_2_motif.png\n');
end

% The demo leaves the toolbox as it found it: the defaults it set at the
% top are restored here.
mptDefaults(prevDefaults);


function [classes, counts, means] = localRank(intervals, density)
%LOCALRANK  Group cells by their interval triple and rank the classes.
%
%   The density at a cell already equals the class's recurrence --- every
%   member of a class sits at the same point, so each sees all c copies.
%   Summing over the members would square that, so the class mean is what
%   recovers the count.
    [classes, ~, idx] = unique(round(intervals).', 'rows');
    counts = accumarray(idx, 1);
    means = accumarray(idx, density(:)) ./ counts;
    [means, order] = sort(means, 'descend');
    classes = classes(order, :);
    counts = counts(order);
end

function s = localTriple(k)
%LOCALTRIPLE  An interval triple as "(3, -3, 5)".
    s = ['(' strjoin(arrayfun(@(v) sprintf('%d', v), k, ...
                              'UniformOutput', false), ', ') ')'];
end

function s = localContour(k)
%LOCALCONTOUR  An interval triple as the scale degrees it traces from 0.
    s = strjoin(arrayfun(@(v) sprintf('%d', v), cumsum([0 k]), ...
                         'UniformOutput', false), char(8594));
end

function localBars(ax, classes, counts, means, top, alsClass, alsColour, ...
                   otherColour, titleText)
%LOCALBARS  The top motifs as a horizontal bar chart, the named class
%   highlighted. A local function cannot see the script's variables, so the
%   count, the class to highlight, and the two colours are passed in.
    sel = top:-1:1;
    yy = 1:top;
    hold(ax, 'on');
    for i = 1:top
        if isequal(classes(sel(i), :), alsClass)
            col = alsColour;
        else
            col = otherColour;
        end
        barh(ax, yy(i), means(sel(i)), 0.56, 'FaceColor', col, ...
             'EdgeColor', 'none');
        text(ax, means(sel(i)), yy(i), ...
             sprintf('  %.1f (%d)', means(sel(i)), counts(sel(i))), ...
             'VerticalAlignment', 'middle', 'FontSize', 12, ...
             'Color', [0.27 0.27 0.27]);
    end
    set(ax, 'YTick', yy, 'YTickLabel', ...
        arrayfun(@(i) localContour(classes(sel(i), :)), 1:top, ...
                 'UniformOutput', false), ...
        'FontSize', 12, 'Box', 'off');
    xlim(ax, [0, max(means(1:top)) * 1.20]);
    ylim(ax, [0.3, top + 0.7]);
    xlabel(ax, 'density (and count)', 'FontSize', 15);
    title(ax, titleText, 'FontSize', 16);
    grid(ax, 'on');
end

function localSlice(ax, gridAxis, field, classes, means, top, alsClass, ...
                    alsColour, titleText)
%LOCALSLICE  A density slice through the leading motif's first interval,
%   with the top classes lying in the slice marked.
    pcolor(ax, gridAxis, gridAxis, field);
    shading(ax, 'interp');
    colormap(ax, jmm.colourMap('magma'));
    cb = colorbar(ax);
    ylabel(cb, 'density', 'FontSize', 13);
    hold(ax, 'on');
    for i = 1:top
        k = classes(i, :);
        if k(1) ~= alsClass(1), continue; end
        als = isequal(k, alsClass);
        if als
            col = alsColour; sz = 130; lw = 2.4;
            label = sprintf('  A Love Supreme  %.1f', means(i));
        else
            col = [1 1 1]; sz = 60; lw = 1.4;
            label = sprintf('  %.1f', means(i));
        end
        scatter(ax, k(2), k(3), sz, 'MarkerEdgeColor', col, 'LineWidth', lw);
        text(ax, k(2), k(3) + 0.35, label, 'Color', col, 'FontSize', 12);
    end
    axis(ax, 'equal');
    xlim(ax, [gridAxis(1), gridAxis(end)]);
    ylim(ax, [gridAxis(1), gridAxis(end)]);
    set(ax, 'XTick', -6:2:6, 'YTick', -6:2:6, 'FontSize', 13, 'Box', 'off');
    xlabel(ax, 'second interval  i_2  (semitones)', 'FontSize', 15);
    ylabel(ax, 'third interval  i_3  (semitones)', 'FontSize', 15);
    title(ax, titleText, 'FontSize', 16);
end

File: python/demos/jmm/demo_jmm_2_2_motif.py
"""demo_jmm_2_2_motif.py — Analysis 2.2 (Online Supplement, Section 8.1):
motif discovery in Acknowledgement from pitch alone.

A demo of the Music Perception Toolbox reproducing the analysis from the
JMM article; lightly edited from the article's own script. Data come
from jmm_data (Acknowledgement read from a MIDI transcription you
supply); the figures stay on screen unless SAVE_FIGURES is set.

Analysis 2.2: data-driven motif discovery in Coltrane's *Acknowledgement*.

The melody is read as one event per note carrying pitch. The recurring
four-note cell is then recovered as a peak in the density of short
interval patterns, without being supplied in advance. Two routes reach
the same transposition-invariant cell.

Differenced route. Pitch is first-differenced to melodic intervals,
collapsing every transposition of a figure onto the same interval
pattern. Three consecutive intervals are then bound into an ordered
super-event: each is a point in a three-dimensional interval space, and
the density over those points is a smoothed recurrence count. A figure
stated k times contributes k near-coincident points, so its interval
triple stands out as a local maximum. Reading the density at every
observed triple and ranking turns motif discovery into peak finding.

Relative route. The same cell is reached without differencing, by
binding four consecutive pitches into an ordered super-event taken
relative: the common transposition is removed, so transposed statements
again coincide, and three degrees of freedom remain — the dimension of
the differenced route's interval triple. The two densities therefore
describe the same object through different internal metrics. The
differenced form treats consecutive intervals independently; the
relative form couples them through the pitch they share. With the
interval kernel set to sqrt(2) times the per-pitch kernel — an interval
being a difference of two pitches, which is the scaling
``difference_events`` applies of its own accord — the two rank the
motifs identically, and differ only in that coupling: a faint shear in
the relative density's slice, absent from the differenced one.

A relative density is read in translation-reduced coordinates: an
r-tuple minus its first value, so a cell's coordinates are its
cumulative intervals. Both the cells and the slice grid are written that
way for the relative route below.

Pre-MAET structure::

    attribute    order  sigma              rel  per
    ----------   -----  -----------------  ---  ---
    dp           3      sqrt(2) * 0.15 st  no   no     differenced route
    pitch        4      0.15 st            yes  no     relative route

    Ordered (exch = 0) in both. Estimator: the density read at each cell.

Data: ``jmm_data.acknowledgement`` (the solo, from your own MIDI
transcription at ``data/AwakeningSolo.mid``). Toolbox:
``pre_maet_from_attr_table``, ``difference_events``, ``bind_events``,
``build_maet``, ``eval_maet``, ``show_pre_maet``. Runtime: a few seconds.
"""
import os
import numpy as np
try:
    import matplotlib.pyplot as plt
except ImportError:
    plt = None
plt.rcParams.update({'font.size': 17, 'axes.titlesize': 19, 'axes.labelsize': 17,
                     'xtick.labelsize': 15, 'ytick.labelsize': 15, 'figure.titlesize': 22,
                     'font.family': 'DejaVu Sans'})

import mpt
_prev_defaults = mpt.set_default(show_hints=False)
from mpt import (pre_maet_from_attr_table, difference_events, bind_events,
                 build_maet, eval_maet, show_pre_maet, unpack_pre_maet)

import jmm_data

# Set True to write the figures to a figures/ folder beside this script;
# False shows them instead.
SAVE_FIGURES = False
FIG_DIR = os.path.join(os.path.dirname(os.path.abspath(__file__)), 'figures')

SIGMA_PITCH = 0.15         # semitones (15 cents): the per-pitch uncertainty
R_DIFF      = 3            # bound interval triples (a four-note cell)
R_REL       = 4            # bound pitch quadruples (the same cell)
TOP         = 6            # motifs shown in the ranking
ALS         = (3, -3, 5)   # +m3, -m3, +P4: the "A Love Supreme" cell

C_ALS   = '#c25008'        # the recurring "A Love Supreme" cell
C_OTHER = '#1f4eb8'        # the surrounding recurring cells


def contour(triple):
    """A cell's interval triple as the scale degrees it traces from 0."""
    return '→'.join(str(int(v)) for v in np.cumsum([0, *triple]))


def rank(intervals, density):
    """Group cells by their interval triple and rank the classes.

    The density at a cell already equals the class's recurrence — every
    member of a class sits at the same point, so each sees all c copies.
    Summing over the members would square that, so the class mean is
    what recovers the count.
    """
    keys, inv = np.unique(np.round(intervals).astype(int), axis=0,
                          return_inverse=True)
    inv = np.asarray(inv).ravel()
    count = np.bincount(inv)
    mean = np.bincount(inv, weights=density) / count
    order = np.argsort(-mean)
    return [(tuple(int(v) for v in keys[i]), int(count[i]), float(mean[i]))
            for i in order]


# --- the melody as a pre-MAET, one event per note -------------------------
notes = jmm_data.acknowledgement()
print(f'{len(notes)} notes; pitch range {notes["pitch"].min():.0f}-'
      f'{notes["pitch"].max():.0f} (MIDI); span '
      f'{notes["onset_beats"].max():.1f} QN')
melody = pre_maet_from_attr_table(
    notes, attributes=(dict(column='pitch', name='pitch', sigma=SIGMA_PITCH),),
    time='beats', chords='separate', weights='ones')

# --- differenced route: intervals bound into ordered triples ---------------
# difference_events widens the kernel by sqrt(2) itself, so the bound
# pre-MAET already carries the interval width and build_maet needs no sigma.
diff_route = bind_events(difference_events(melody, 1), R_DIFF, step=1)
show_pre_maet(diff_route, max_events=3)
cells_d = unpack_pre_maet(diff_route)[0][0]              # (3, nCells) intervals
dens_d = np.asarray(eval_maet(build_maet(diff_route, verbose=False),
                              cells_d, verbose=False))
ranked_d = rank(cells_d.T, dens_d)

# --- relative route: pitches bound into ordered relative quadruples -------
rel_route = bind_events(melody, R_REL, step=1, rel_outer=True)
show_pre_maet(rel_route, max_events=3)
cells_r = unpack_pre_maet(rel_route)[0][0]               # (4, nCells) pitches
iv_r = np.diff(cells_r, axis=0)                          # its interval triples
dens_r = np.asarray(eval_maet(build_maet(rel_route, verbose=False),
                              np.cumsum(iv_r, axis=0), verbose=False))
ranked_r = rank(iv_r.T, dens_r)

# --- report ---------------------------------------------------------------
print(f'\ndifferenced: {cells_d.shape[1]} ordered interval triples, '
      f'{len(ranked_d)} distinct classes')
print(f'relative:    {cells_r.shape[1]} ordered pitch quadruples, '
      f'{len(ranked_r)} distinct classes\n')
print(f'{"rank":>4}  {"interval class":>16}  {"count":>5}  '
      f'{"differenced":>11}  {"relative":>9}   contour')
by_class = {k: d for k, _, d in ranked_r}
for i, (k, c, d) in enumerate(ranked_d[:TOP], 1):
    mark = '  <- A Love Supreme cell' if k == ALS else ''
    print(f'{i:>4}  {str(k):>16}  {c:>5}  {d:>11.1f}  {by_class[k]:>9.1f}   '
          f'{contour(k)}{mark}')
print(f'\ntop-{TOP + 2} interval-class set identical: '
      f'{set(k for k, _, _ in ranked_d[:TOP + 2]) == set(k for k, _, _ in ranked_r[:TOP + 2])}')
print(f'A Love Supreme cell {ALS}: rank '
      f'{1 + [k for k, _, _ in ranked_d].index(ALS)} (differenced), rank '
      f'{1 + [k for k, _, _ in ranked_r].index(ALS)} (relative)')

# --- the (+3, i2, i3) slice of each density -------------------------------
# The plane of cells sharing the leading motif's first interval. The grid
# is written as interval triples for the differenced route and as their
# cumulative sums for the relative one.
step = SIGMA_PITCH * np.sqrt(2.0) / 3.0
axis = np.arange(-7.0, 7.0 + 0.5 * step, step)
I2, I3 = np.meshgrid(axis, axis)
grid_iv = np.vstack([np.full(I2.size, float(ALS[0])), I2.ravel(), I3.ravel()])
slice_d = np.asarray(eval_maet(build_maet(diff_route, verbose=False),
                               grid_iv, verbose=False)).reshape(I2.shape)
slice_r = np.asarray(eval_maet(build_maet(rel_route, verbose=False),
                               np.cumsum(grid_iv, axis=0),
                               verbose=False)).reshape(I2.shape)
print(f'\nslice peak (i2, i3): differenced '
      f'({axis[np.unravel_index(slice_d.argmax(), slice_d.shape)[1]]:.2f}, '
      f'{axis[np.unravel_index(slice_d.argmax(), slice_d.shape)[0]]:.2f}), '
      f'relative '
      f'({axis[np.unravel_index(slice_r.argmax(), slice_r.shape)[1]]:.2f}, '
      f'{axis[np.unravel_index(slice_r.argmax(), slice_r.shape)[0]]:.2f})')

# --- figure ---------------------------------------------------------------
if plt is None:
    print('matplotlib not available; skipping the figure.')
    mpt.set_default(**_prev_defaults)
    raise SystemExit(0)

fig, axes = plt.subplots(2, 2, figsize=(15, 10.7),
                         gridspec_kw={'height_ratios': [3.7, 4.06]})


def draw_bars(ax, ranked, title):
    top = ranked[:TOP][::-1]
    yy = np.arange(len(top))
    ax.barh(yy, [d for _, _, d in top], height=0.56,
            color=[C_ALS if k == ALS else C_OTHER for k, _, _ in top])
    ax.set_yticks(yy)
    ax.set_yticklabels([contour(k) for k, _, _ in top], fontsize=13)
    for y, (_, c, d) in zip(yy, top):
        ax.text(d, y, f'  {d:.1f} ({c})', va='center', fontsize=14, color='#444')
    ax.set_xlabel('density (and count)')
    ax.set_title(title)
    ax.spines[['top', 'right']].set_visible(False)
    ax.grid(True, axis='x', alpha=0.3)
    ax.set_xlim(0, max(d for _, _, d in top) * 1.20)
    ax.set_ylim(-0.7, len(top) - 0.3)


def draw_slice(ax, field, ranked, title):
    pc = ax.pcolormesh(axis, axis, field, cmap='magma', shading='gouraud')
    pc.set_rasterized(True)
    fig.colorbar(pc, ax=ax, fraction=0.046, pad=0.04, label='density')
    for k, _, d in ranked[:TOP]:
        if k[0] != ALS[0]:
            continue
        als = k == ALS
        ax.scatter(k[1], k[2], s=130 if als else 60, facecolors='none',
                   edgecolors=C_ALS if als else 'white',
                   linewidths=2.4 if als else 1.4, zorder=3)
        ax.annotate(('A Love Supreme  ' if als else '') + f'{d:.1f}',
                    (k[1], k[2]), textcoords='offset points', xytext=(7, 6),
                    fontsize=14, color=C_ALS if als else 'white',
                    fontweight='bold' if als else 'normal')
    ax.set_xlabel('second interval  $i_2$  (semitones)')
    ax.set_ylabel('third interval  $i_3$  (semitones)')
    ax.set_xticks(np.arange(-6, 7, 2))
    ax.set_yticks(np.arange(-6, 7, 2))
    ax.set_title(title)
    ax.set_aspect('equal')


draw_bars(axes[0, 0], ranked_d, 'Differenced interval triples')
draw_bars(axes[0, 1], ranked_r, 'Relative pitch quadruples')
draw_slice(axes[1, 0], slice_d, ranked_d,
           fr'Density through $(+{ALS[0]},\, i_2,\, i_3)$: differenced')
draw_slice(axes[1, 1], slice_r, ranked_r,
           fr'Density through $(+{ALS[0]},\, i_2,\, i_3)$: relative')
fig.suptitle('Coltrane, Acknowledgement: motif density', y=0.985, fontsize=15)
fig.subplots_adjust(left=0.115, right=0.93, top=0.90, bottom=0.075,
                    hspace=0.275, wspace=0.40)

if SAVE_FIGURES:
    os.makedirs(FIG_DIR, exist_ok=True)
    fig.savefig(os.path.join(FIG_DIR, 'demo_jmm_2_2_motif.png'),
                dpi=140, bbox_inches='tight')
    print('Saved figures/demo_jmm_2_2_motif.png')
else:
    plt.show()

# The demo leaves the toolbox as it found it: the defaults it set at the
# top are restored here.
mpt.set_default(**_prev_defaults)
